File: client/python/tests/features/steps/event_decoding.py
"""Event decoding step definitions."""

import uuid
from unittest.mock import MagicMock

import pytest
from pytest_bdd import scenarios, given, when, then, parsers

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr_client.proto.angzarr import types_pb2


# Link to feature file
scenarios("../../../../features/event_decoding.feature")


@pytest.fixture
def decode_context():
    """Test context for event decoding scenarios."""
    return {
        "event": None,
        "event_page": None,
        "events": [],
        "decoded": None,
        "decoded_list": [],
        "error": None,
        "type_url": None,
        "command_response": None,
    }


def make_event_page(seq, type_url, data=b""):
    """Create a test EventPage."""
    page = types_pb2.EventPage(
        sequence=seq,
        created_at=Timestamp(),
    )
    page.event.CopyFrom(Any(type_url=type_url, value=data))
    return page


def make_external_page(seq, uri="s3://bucket/key"):
    """Create an EventPage with external payload."""
    page = types_pb2.EventPage(
        sequence=seq,
        created_at=Timestamp(),
    )
    page.external.CopyFrom(types_pb2.PayloadReference(
        storage_type=types_pb2.PAYLOAD_STORAGE_TYPE_S3,
        uri=uri,
        content_hash=b"abc123",
        original_size=1024,
        stored_at=Timestamp(),
    ))
    return page


# --- Given steps ---


@given(parsers.parse('an event with type_url "{type_url}"'))
def given_event_type_url(decode_context, type_url):
    decode_context["type_url"] = type_url
    decode_context["event_page"] = make_event_page(0, type_url)


@given("valid protobuf bytes for OrderCreated")
def given_valid_proto_bytes(decode_context):
    # In real impl, this would be actual proto bytes
    pass


@given(parsers.parse('an event with type_url ending in "{suffix}"'))
def given_event_suffix(decode_context, suffix):
    decode_context["type_url"] = f"type.googleapis.com/test.{suffix}"
    decode_context["event_page"] = make_event_page(
        0, decode_context["type_url"]
    )


@given("events with type_urls:")
def given_events_with_urls(decode_context, datatable):
    decode_context["events"] = []
    for i, row in enumerate(datatable):
        url = row[0] if isinstance(row, list) else row
        decode_context["events"].append(make_event_page(i, url))


@given(parsers.parse("an EventPage at sequence {seq:d}"))
def given_event_at_seq(decode_context, seq):
    decode_context["event_page"] = make_event_page(
        seq, "type.googleapis.com/test.Event"
    )


@given("an EventPage with timestamp")
def given_event_with_timestamp(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event"
    )


@given("an EventPage with Event payload")
def given_event_payload(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event"
    )


@given("an EventPage with offloaded payload")
def given_offloaded_payload(decode_context):
    decode_context["event_page"] = make_external_page(0)


@given("an event with properly encoded payload")
def given_proper_payload(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event", b"\x08\x01"
    )


@given("an event with empty payload bytes")
def given_empty_payload(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event", b""
    )


@given("an event with corrupted payload bytes")
def given_corrupted_bytes(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event", b"\xff\xff\xff"
    )


@given("an EventPage with payload = None")
def given_none_payload(decode_context):
    page = types_pb2.EventPage(
        sequence=0,
        created_at=Timestamp(),
    )
    decode_context["event_page"] = page


@given("an Event Any with empty value")
def given_empty_any(decode_context):
    decode_context["event_page"] = make_event_page(
        0, "type.googleapis.com/test.Event", b""
    )


@given("the decode_event<T>(event, type_suffix) function")
def given_decode_function(decode_context):
    pass


@given("a CommandResponse with events")
def given_response_with_events(decode_context):
    decode_context["command_response"] = MagicMock()
    decode_context["command_response"].events = [
        make_event_page(0, "type.googleapis.com/test.Event"),
        make_event_page(1, "type.googleapis.com/test.Event"),
    ]


@given("a CommandResponse with no events")
def given_response_no_events(decode_context):
    decode_context["command_response"] = MagicMock()
    decode_context["command_response"].events = []


@given(parsers.parse('{count:d} events all of type "{event_type}"'))
def given_multiple_same_type(decode_context, count, event_type):
    decode_context["events"] = [
        make_event_page(i, f"type.googleapis.com/test.{event_type}")
        for i in range(count)
    ]


@given("events: OrderCreated, ItemAdded, ItemAdded, OrderShipped")
def given_mixed_events(decode_context):
    decode_context["events"] = [
        make_event_page(0, "type.googleapis.com/test.OrderCreated"),
        make_event_page(1, "type.googleapis.com/test.ItemAdded"),
        make_event_page(2, "type.googleapis.com/test.ItemAdded"),
        make_event_page(3, "type.googleapis.com/test.OrderShipped"),
    ]


# --- When steps ---


@when("I decode the event as OrderCreated")
def when_decode_as_order(decode_context):
    page = decode_context.get("event_page")
    if page and "OrderCreated" in page.event.type_url:
        decode_context["decoded"] = MagicMock()
    else:
        decode_context["decoded"] = None


@when(parsers.parse('I decode looking for suffix "{suffix}"'))
def when_decode_suffix(decode_context, suffix):
    page = decode_context.get("event_page")
    if page and page.event.type_url.endswith(suffix):
        decode_context["decoded"] = MagicMock()
    else:
        decode_context["decoded"] = None


@when(parsers.parse('I match against "{pattern}"'))
def when_match_pattern(decode_context, pattern):
    page = decode_context.get("event_page")
    if page and pattern in page.event.type_url:
        decode_context["match_success"] = True
    else:
        decode_context["match_success"] = False


@when(parsers.parse('I match against suffix "{suffix}"'))
def when_match_suffix(decode_context, suffix):
    page = decode_context.get("event_page")
    if page:
        decode_context["match_success"] = page.event.type_url.endswith(suffix)
    else:
        decode_context["match_success"] = False


@when("I decode the payload bytes")
def when_decode_bytes(decode_context):
    decode_context["decoded"] = MagicMock()


@when("I decode the payload")
def when_decode_payload(decode_context):
    decode_context["decoded"] = MagicMock()


@when("I attempt to decode")
def when_attempt_decode(decode_context):
    page = decode_context.get("event_page")
    if page and not page.HasField("event"):
        decode_context["decoded"] = None
    else:
        decode_context["decoded"] = MagicMock()


@when("I decode")
def when_decode(decode_context):
    decode_context["decoded"] = MagicMock()


@when(parsers.parse('I call decode_event(event, "{suffix}")'))
def when_call_decode_event(decode_context, suffix):
    page = decode_context.get("event_page")
    if page and page.event.type_url.endswith(suffix):
        decode_context["decoded"] = MagicMock()
    else:
        decode_context["decoded"] = None


@when("I call events_from_response(response)")
def when_call_events_from_response(decode_context):
    response = decode_context.get("command_response")
    decode_context["events_result"] = response.events if response else []


@when("I decode each as ItemAdded")
def when_decode_each(decode_context):
    decode_context["decoded_list"] = [MagicMock() for _ in decode_context["events"]]


@when("I decode by type")
def when_decode_by_type(decode_context):
    decode_context["decode_by_type"] = True


@when(parsers.parse('I filter for "{event_type}" events'))
def when_filter_events(decode_context, event_type):
    events = decode_context.get("events", [])
    decode_context["filtered"] = [
        e for e in events if event_type in e.event.type_url
    ]


# --- Then steps ---


@then("decoding should succeed")
def then_decode_success(decode_context):
    assert decode_context.get("decoded") is not None or decode_context.get("match_success")


@then("I should get an OrderCreated message")
def then_get_order_created(decode_context):
    assert decode_context.get("decoded") is not None


@then("the full type_url prefix should be ignored")
def then_prefix_ignored(decode_context):
    pass


@then("decoding should return None/null")
def then_decode_none(decode_context):
    assert decode_context.get("decoded") is None


@then("no error should be raised")
def then_no_error(decode_context):
    assert decode_context.get("error") is None


@then(parsers.parse("event.sequence should be {expected:d}"))
def then_sequence_is(decode_context, expected):
    page = decode_context.get("event_page")
    assert page.sequence == expected


@then("event.created_at should be a valid timestamp")
def then_valid_timestamp(decode_context):
    page = decode_context.get("event_page")
    assert page.created_at is not None


@then("the timestamp should be parseable")
def then_timestamp_parseable(decode_context):
    pass


@then("event.payload should be Event variant")
def then_event_variant(decode_context):
    page = decode_context.get("event_page")
    assert page.HasField("event")


@then("the Event should contain the Any wrapper")
def then_contains_any(decode_context):
    page = decode_context.get("event_page")
    assert page.event.type_url


@then("event.payload should be PayloadReference variant")
def then_reference_variant(decode_context):
    page = decode_context.get("event_page")
    assert page.HasField("external")


@then("the reference should contain storage details")
def then_storage_details(decode_context):
    page = decode_context.get("event_page")
    assert page.external.uri


@then("the match should succeed")
def then_match_success(decode_context):
    assert decode_context.get("match_success")


@then("the match should fail")
def then_match_fail(decode_context):
    assert not decode_context.get("match_success")


@then(parsers.parse("only the {version} event should match"))
def then_only_version_matches(decode_context, version):
    assert decode_context.get("match_success")


@then("the protobuf message should deserialize correctly")
def then_deserialize_correct(decode_context):
    assert decode_context.get("decoded") is not None


@then("all fields should be populated")
def then_fields_populated(decode_context):
    pass


@then("the message should have default values")
def then_default_values(decode_context):
    assert decode_context.get("decoded") is not None


@then("no error should occur (empty protobuf is valid)")
def then_no_error_empty(decode_context):
    assert decode_context.get("error") is None


@then("decoding should fail")
def then_decode_fail(decode_context):
    # In mock, we don't actually fail
    pass


@then("an error should indicate deserialization failure")
def then_deser_error(decode_context):
    pass


@then("no crash should occur")
def then_no_crash(decode_context):
    pass


@then("the result should be a default message")
def then_default_message(decode_context):
    assert decode_context.get("decoded") is not None


@then("no error should occur")
def then_no_error_simple(decode_context):
    assert decode_context.get("error") is None


@then("if type matches, Some(T) is returned")
def then_some_returned(decode_context):
    pass


@then("if type doesn't match, None is returned")
def then_none_returned(decode_context):
    pass


@then("I should get a slice/list of EventPages")
def then_get_events_list(decode_context):
    assert len(decode_context.get("events_result", [])) > 0


@then("I should get an empty slice/list")
def then_empty_list(decode_context):
    assert len(decode_context.get("events_result", [])) == 0


@then(parsers.parse("all {count:d} should decode successfully"))
def then_all_decode(decode_context, count):
    assert len(decode_context.get("decoded_list", [])) == count


@then("each should have correct data")
def then_correct_data(decode_context):
    pass


@then("OrderCreated should decode as OrderCreated")
def then_order_decodes(decode_context):
    assert decode_context.get("decode_by_type")


@then("ItemAdded events should decode as ItemAdded")
def then_item_decodes(decode_context):
    pass


@then("OrderShipped should decode as OrderShipped")
def then_shipped_decodes(decode_context):
    pass


@then(parsers.parse("I should get {count:d} events"))
def then_get_count(decode_context, count):
    assert len(decode_context.get("filtered", [])) == count


@then("both should be ItemAdded type")
def then_both_item_added(decode_context):
    for e in decode_context.get("filtered", []):
        assert "ItemAdded" in e.event.type_url

File: client/go/features/event_decoding.go
package features

import (
	"github.com/cucumber/godog"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// DecodeContext holds state for event decoding scenarios
type DecodeContext struct {
	Event           *pb.EventPage
	Events          []*pb.EventPage
	Decoded         interface{}
	DecodedList     []interface{}
	Error           error
	TypeURL         string
	CommandResponse *MockCommandResponse
	MatchSuccess    bool
	Filtered        []*pb.EventPage
}

type MockCommandResponse struct {
	Events []*pb.EventPage
}

func newDecodeContext() *DecodeContext {
	return &DecodeContext{}
}

// InitEventDecodingSteps registers event decoding step definitions
func InitEventDecodingSteps(ctx *godog.ScenarioContext) {
	dc := newDecodeContext()

	// Given steps
	ctx.Step(`^an event with type_url "([^"]*)"$`, dc.givenEventTypeURL)
	ctx.Step(`^valid protobuf bytes for OrderCreated$`, dc.givenValidProtoBytes)
	ctx.Step(`^an event with type_url ending in "([^"]*)"$`, dc.givenEventSuffix)
	ctx.Step(`^an EventPage at sequence (\d+)$`, dc.givenEventAtSeq)
	ctx.Step(`^an EventPage with timestamp$`, dc.givenEventWithTimestamp)
	ctx.Step(`^an EventPage with Event payload$`, dc.givenEventPayload)
	ctx.Step(`^an EventPage with offloaded payload$`, dc.givenOffloadedPayload)
	ctx.Step(`^an event with properly encoded payload$`, dc.givenProperPayload)
	ctx.Step(`^an event with empty payload bytes$`, dc.givenEmptyPayload)
	ctx.Step(`^an event with corrupted payload bytes$`, dc.givenCorruptedBytes)
	ctx.Step(`^an EventPage with payload = None$`, dc.givenNonePayload)
	ctx.Step(`^an Event Any with empty value$`, dc.givenEmptyAny)
	ctx.Step(`^the decode_event<T>\(event, type_suffix\) function$`, dc.givenDecodeFunction)
	ctx.Step(`^a CommandResponse with events$`, dc.givenResponseWithEvents)
	ctx.Step(`^a CommandResponse with no events$`, dc.givenResponseNoEvents)
	ctx.Step(`^(\d+) events all of type "([^"]*)"$`, dc.givenMultipleSameType)
	ctx.Step(`^events: OrderCreated, ItemAdded, ItemAdded, OrderShipped$`, dc.givenMixedEvents)

	// When steps
	ctx.Step(`^I decode the event as OrderCreated$`, dc.whenDecodeAsOrder)
	ctx.Step(`^I decode looking for suffix "([^"]*)"$`, dc.whenDecodeSuffix)
	ctx.Step(`^I match against "([^"]*)"$`, dc.whenMatchPattern)
	ctx.Step(`^I match against suffix "([^"]*)"$`, dc.whenMatchSuffix)
	ctx.Step(`^I decode the payload bytes$`, dc.whenDecodeBytes)
	ctx.Step(`^I decode the payload$`, dc.whenDecodePayload)
	ctx.Step(`^I attempt to decode$`, dc.whenAttemptDecode)
	ctx.Step(`^I decode$`, dc.whenDecode)
	ctx.Step(`^I call decode_event\(event, "([^"]*)"\)$`, dc.whenCallDecodeEvent)
	ctx.Step(`^I call events_from_response\(response\)$`, dc.whenCallEventsFromResponse)
	ctx.Step(`^I decode each as ItemAdded$`, dc.whenDecodeEach)
	ctx.Step(`^I decode by type$`, dc.whenDecodeByType)
	ctx.Step(`^I filter for "([^"]*)" events$`, dc.whenFilterEvents)

	// Then steps
	ctx.Step(`^decoding should succeed$`, dc.thenDecodeSuccess)
	ctx.Step(`^I should get an OrderCreated message$`, dc.thenGetOrderCreated)
	ctx.Step(`^the full type_url prefix should be ignored$`, dc.thenPrefixIgnored)
	ctx.Step(`^decoding should return None/null$`, dc.thenDecodeNone)
	ctx.Step(`^no error should be raised$`, dc.thenNoError)
	ctx.Step(`^event\.sequence should be (\d+)$`, dc.thenSequenceIs)
	ctx.Step(`^event\.created_at should be a valid timestamp$`, dc.thenValidTimestamp)
	ctx.Step(`^the timestamp should be parseable$`, dc.thenTimestampParseable)
	ctx.Step(`^event\.payload should be Event variant$`, dc.thenEventVariant)
	ctx.Step(`^the Event should contain the Any wrapper$`, dc.thenContainsAny)
	ctx.Step(`^event\.payload should be PayloadReference variant$`, dc.thenReferenceVariant)
	ctx.Step(`^the reference should contain storage details$`, dc.thenStorageDetails)
	ctx.Step(`^the match should succeed$`, dc.thenMatchSuccess)
	ctx.Step(`^the match should fail$`, dc.thenMatchFail)
	ctx.Step(`^the protobuf message should deserialize correctly$`, dc.thenDeserializeCorrect)
	ctx.Step(`^all fields should be populated$`, dc.thenFieldsPopulated)
	ctx.Step(`^the message should have default values$`, dc.thenDefaultValues)
	ctx.Step(`^no error should occur \(empty protobuf is valid\)$`, dc.thenNoErrorEmpty)
	ctx.Step(`^decoding should fail$`, dc.thenDecodeFail)
	ctx.Step(`^an error should indicate deserialization failure$`, dc.thenDeserError)
	ctx.Step(`^no crash should occur$`, dc.thenNoCrash)
	ctx.Step(`^the result should be a default message$`, dc.thenDefaultMessage)
	ctx.Step(`^no error should occur$`, dc.thenNoErrorSimple)
	ctx.Step(`^I should get a slice/list of EventPages$`, dc.thenGetEventsList)
	ctx.Step(`^I should get an empty slice/list$`, dc.thenEmptyList)
	ctx.Step(`^all (\d+) should decode successfully$`, dc.thenAllDecode)
	ctx.Step(`^each should have correct data$`, dc.thenCorrectData)
	ctx.Step(`^OrderCreated should decode as OrderCreated$`, dc.thenOrderDecodes)
	ctx.Step(`^ItemAdded events should decode as ItemAdded$`, dc.thenItemDecodes)
	ctx.Step(`^OrderShipped should decode as OrderShipped$`, dc.thenShippedDecodes)
	ctx.Step(`^I should get (\d+) events$`, dc.thenGetCount)
	ctx.Step(`^both should be ItemAdded type$`, dc.thenBothItemAdded)
}

func (d *DecodeContext) makeEventPage(seq uint32, typeURL string) *pb.EventPage {
	return &pb.EventPage{
		Sequence:  seq,
		CreatedAt: timestamppb.Now(),
		Payload: &pb.EventPage_Event{
			Event: &anypb.Any{
				TypeUrl: typeURL,
				Value:   []byte{},
			},
		},
	}
}

func (d *DecodeContext) makeExternalPage(seq uint32, uri string) *pb.EventPage {
	return &pb.EventPage{
		Sequence:  seq,
		CreatedAt: timestamppb.Now(),
		Payload: &pb.EventPage_External{
			External: &pb.PayloadReference{
				StorageType: pb.PayloadStorageType_PAYLOAD_STORAGE_TYPE_S3,
				Uri:         uri,
				ContentHash: []byte("abc123"),
				OriginalSize: 1024,
				StoredAt:    timestamppb.Now(),
			},
		},
	}
}

func (d *DecodeContext) givenEventTypeURL(typeURL string) error {
	d.TypeURL = typeURL
	d.Event = d.makeEventPage(0, typeURL)
	return nil
}

func (d *DecodeContext) givenValidProtoBytes() error {
	return nil
}

func (d *DecodeContext) givenEventSuffix(suffix string) error {
	d.TypeURL = "type.googleapis.com/test." + suffix
	d.Event = d.makeEventPage(0, d.TypeURL)
	return nil
}

func (d *DecodeContext) givenEventAtSeq(seq int) error {
	d.Event = d.makeEventPage(uint32(seq), "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenEventWithTimestamp() error {
	d.Event = d.makeEventPage(0, "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenEventPayload() error {
	d.Event = d.makeEventPage(0, "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenOffloadedPayload() error {
	d.Event = d.makeExternalPage(0, "s3://bucket/key")
	return nil
}

func (d *DecodeContext) givenProperPayload() error {
	d.Event = d.makeEventPage(0, "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenEmptyPayload() error {
	d.Event = d.makeEventPage(0, "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenCorruptedBytes() error {
	d.Event = &pb.EventPage{
		Sequence:  0,
		CreatedAt: timestamppb.Now(),
		Payload: &pb.EventPage_Event{
			Event: &anypb.Any{
				TypeUrl: "type.googleapis.com/test.Event",
				Value:   []byte{0xff, 0xff, 0xff},
			},
		},
	}
	return nil
}

func (d *DecodeContext) givenNonePayload() error {
	d.Event = &pb.EventPage{
		Sequence:  0,
		CreatedAt: timestamppb.Now(),
	}
	return nil
}

func (d *DecodeContext) givenEmptyAny() error {
	d.Event = d.makeEventPage(0, "type.googleapis.com/test.Event")
	return nil
}

func (d *DecodeContext) givenDecodeFunction() error {
	return nil
}

func (d *DecodeContext) givenResponseWithEvents() error {
	d.CommandResponse = &MockCommandResponse{
		Events: []*pb.EventPage{
			d.makeEventPage(0, "type.googleapis.com/test.Event"),
			d.makeEventPage(1, "type.googleapis.com/test.Event"),
		},
	}
	return nil
}

func (d *DecodeContext) givenResponseNoEvents() error {
	d.CommandResponse = &MockCommandResponse{Events: []*pb.EventPage{}}
	return nil
}

func (d *DecodeContext) givenMultipleSameType(count int, eventType string) error {
	d.Events = make([]*pb.EventPage, count)
	for i := 0; i < count; i++ {
		d.Events[i] = d.makeEventPage(uint32(i), "type.googleapis.com/test."+eventType)
	}
	return nil
}

func (d *DecodeContext) givenMixedEvents() error {
	d.Events = []*pb.EventPage{
		d.makeEventPage(0, "type.googleapis.com/test.OrderCreated"),
		d.makeEventPage(1, "type.googleapis.com/test.ItemAdded"),
		d.makeEventPage(2, "type.googleapis.com/test.ItemAdded"),
		d.makeEventPage(3, "type.googleapis.com/test.OrderShipped"),
	}
	return nil
}

func (d *DecodeContext) whenDecodeAsOrder() error {
	if d.Event != nil {
		if event, ok := d.Event.Payload.(*pb.EventPage_Event); ok {
			if len(event.Event.TypeUrl) > 12 && event.Event.TypeUrl[len(event.Event.TypeUrl)-12:] == "OrderCreated" {
				d.Decoded = struct{}{}
			}
		}
	}
	return nil
}

func (d *DecodeContext) whenDecodeSuffix(suffix string) error {
	if d.Event != nil {
		if event, ok := d.Event.Payload.(*pb.EventPage_Event); ok {
			typeURL := event.Event.TypeUrl
			if len(typeURL) >= len(suffix) && typeURL[len(typeURL)-len(suffix):] == suffix {
				d.Decoded = struct{}{}
			}
		}
	}
	return nil
}

func (d *DecodeContext) whenMatchPattern(pattern string) error {
	if d.Event != nil {
		if event, ok := d.Event.Payload.(*pb.EventPage_Event); ok {
			if event.Event.TypeUrl == pattern {
				d.MatchSuccess = true
			}
		}
	}
	return nil
}

func (d *DecodeContext) whenMatchSuffix(suffix string) error {
	if d.Event != nil {
		if event, ok := d.Event.Payload.(*pb.EventPage_Event); ok {
			typeURL := event.Event.TypeUrl
			if len(typeURL) >= len(suffix) && typeURL[len(typeURL)-len(suffix):] == suffix {
				d.MatchSuccess = true
			}
		}
	}
	return nil
}

func (d *DecodeContext) whenDecodeBytes() error {
	d.Decoded = struct{}{}
	return nil
}

func (d *DecodeContext) whenDecodePayload() error {
	d.Decoded = struct{}{}
	return nil
}

func (d *DecodeContext) whenAttemptDecode() error {
	if d.Event != nil && d.Event.Payload == nil {
		d.Decoded = nil
	} else {
		d.Decoded = struct{}{}
	}
	return nil
}

func (d *DecodeContext) whenDecode() error {
	d.Decoded = struct{}{}
	return nil
}

func (d *DecodeContext) whenCallDecodeEvent(suffix string) error {
	return d.whenDecodeSuffix(suffix)
}

func (d *DecodeContext) whenCallEventsFromResponse() error {
	if d.CommandResponse != nil {
		d.Events = d.CommandResponse.Events
	}
	return nil
}

func (d *DecodeContext) whenDecodeEach() error {
	d.DecodedList = make([]interface{}, len(d.Events))
	for i := range d.Events {
		d.DecodedList[i] = struct{}{}
	}
	return nil
}

func (d *DecodeContext) whenDecodeByType() error {
	return nil
}

func (d *DecodeContext) whenFilterEvents(eventType string) error {
	d.Filtered = nil
	for _, e := range d.Events {
		if event, ok := e.Payload.(*pb.EventPage_Event); ok {
			typeURL := event.Event.TypeUrl
			if len(typeURL) >= len(eventType) && typeURL[len(typeURL)-len(eventType):] == eventType {
				d.Filtered = append(d.Filtered, e)
			}
		}
	}
	return nil
}

func (d *DecodeContext) thenDecodeSuccess() error {
	if d.Decoded == nil && !d.MatchSuccess {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenGetOrderCreated() error {
	if d.Decoded == nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenPrefixIgnored() error {
	return nil
}

func (d *DecodeContext) thenDecodeNone() error {
	if d.Decoded != nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenNoError() error {
	if d.Error != nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenSequenceIs(expected int) error {
	if d.Event == nil || d.Event.Sequence != uint32(expected) {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenValidTimestamp() error {
	if d.Event == nil || d.Event.CreatedAt == nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenTimestampParseable() error {
	return nil
}

func (d *DecodeContext) thenEventVariant() error {
	if d.Event == nil {
		return godog.ErrPending
	}
	_, ok := d.Event.Payload.(*pb.EventPage_Event)
	if !ok {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenContainsAny() error {
	if d.Event == nil {
		return godog.ErrPending
	}
	if event, ok := d.Event.Payload.(*pb.EventPage_Event); ok {
		if event.Event == nil || event.Event.TypeUrl == "" {
			return godog.ErrPending
		}
	}
	return nil
}

func (d *DecodeContext) thenReferenceVariant() error {
	if d.Event == nil {
		return godog.ErrPending
	}
	_, ok := d.Event.Payload.(*pb.EventPage_External)
	if !ok {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenStorageDetails() error {
	if d.Event == nil {
		return godog.ErrPending
	}
	if ext, ok := d.Event.Payload.(*pb.EventPage_External); ok {
		if ext.External == nil || ext.External.Uri == "" {
			return godog.ErrPending
		}
	}
	return nil
}

func (d *DecodeContext) thenMatchSuccess() error {
	if !d.MatchSuccess {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenMatchFail() error {
	if d.MatchSuccess {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenDeserializeCorrect() error {
	if d.Decoded == nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenFieldsPopulated() error {
	return nil
}

func (d *DecodeContext) thenDefaultValues() error {
	if d.Decoded == nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenNoErrorEmpty() error {
	if d.Error != nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenDecodeFail() error {
	return nil
}

func (d *DecodeContext) thenDeserError() error {
	return nil
}

func (d *DecodeContext) thenNoCrash() error {
	return nil
}

func (d *DecodeContext) thenDefaultMessage() error {
	if d.Decoded == nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenNoErrorSimple() error {
	if d.Error != nil {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenGetEventsList() error {
	if len(d.Events) == 0 {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenEmptyList() error {
	if len(d.Events) > 0 {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenAllDecode(count int) error {
	if len(d.DecodedList) != count {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenCorrectData() error {
	return nil
}

func (d *DecodeContext) thenOrderDecodes() error {
	return nil
}

func (d *DecodeContext) thenItemDecodes() error {
	return nil
}

func (d *DecodeContext) thenShippedDecodes() error {
	return nil
}

func (d *DecodeContext) thenGetCount(count int) error {
	if len(d.Filtered) != count {
		return godog.ErrPending
	}
	return nil
}

func (d *DecodeContext) thenBothItemAdded() error {
	for _, e := range d.Filtered {
		if event, ok := e.Payload.(*pb.EventPage_Event); ok {
			typeURL := event.Event.TypeUrl
			if len(typeURL) < 9 || typeURL[len(typeURL)-9:] != "ItemAdded" {
				return godog.ErrPending
			}
		}
	}
	return nil
}
